docs/types.rst

```diff
@@ -805,40 +805,48 @@ if the local variable itself is just a reference.
 Arrays
 ------
 
-Arrays can have a compile-time fixed size or they can be dynamic.
-The are few restrictions for the element, it can also be
-another array, a mapping or a struct. The general restrictions for
-types apply, though, in that mappings can only be used in storage
-and publicly-visible functions need parameters that are :ref:`ABI types <ABI>`.
+Arrays can have a compile-time fixed size, or they can have a dynamic size.
 
 An array of fixed size ``k`` and element type ``T`` is written as ``T[k]``,
-an array of dynamic size as ``T[]``. As an example, an array of 5 dynamic
-arrays of ``uint`` is ``uint[][5]`` (note that the notation is reversed when
-compared to some other languages). To access the second uint in the
-third dynamic array, you use ``x[2][1]`` (indices are zero-based and
-access works in the opposite way of the declaration, i.e. ``x[2]``
-shaves off one level in the type from the right).
+and an array of dynamic size as ``T[]``.
 
-Accessing an array past its end causes a revert. If you want to add
-new elements, you have to use ``.push()`` or increase the ``.length``
-member (see below).
+For example, an array of 5 dynamic arrays of ``uint`` is written as
+``uint[][5]``. The notation is reversed compared to some other languages, as in
+Solidity, ``X[3]`` is always an array containing three elements of type ``X``,
+even if ``X`` is itself an array. This is not the case in other languages such
+as C.
+
+Indices are zero-based, and access is in the opposite direction of the
+declaration.
+For example to access the second ``uint`` in the third dynamic array, use
+``unit[2][1]``, and to access the third dynamic array, use ``unit[2]``. Again,
+if you have an array ``T[5] a`` for a type ``T`` that can also be an array,
+then ``a[2]`` always has type ``T``.
+
+Array elements can be of any type, including mapping or struct. The general
+restrictions for types apply, in that mappings can only be stored in the
```

Comment:
"stored in storage"

Reply:
@chriseth Yeah, not too happy about it, but it's the word we've been using so far and I think consistency trumps the dodgy grammar. I code formatted ``storage`` to make it a little better.

Reply:
My point is that "data location" is just the thing attached a type name. Here we also just call it "storage": https://solidity.readthedocs.io/en/v0.5.1/introduction-to-smart-contracts.html#storage-memory-and-the-stack

Reply:
@chriseth OK, I understand, you mean use 

"stored in storage" instead of "stored in the
``storage`` data location", I missed that from the comment.

OK, I am not opposed to either description, but at some point in the past few months we decided to standardise on "``x`` data location", but of course, this was in progress of change as relevant PRs were submitted. All I need to know is which one to stick with from now on and that's fine and I can correct references to the one style.

Reply:
"In `StructName storage x` the variable `x` has a data location of `storage` and thus its data resides in storage."